File: neetcode-250/trees/543-diameter-of-binary-tree.ts
import { TreeNode } from "./treeNode";

function diameterOfBinaryTree(root: TreeNode | null): number {
	let res = 0;
	dfs(root);
	return res;

	function dfs(node: TreeNode | null): number {
		if (!node) {
			return 0;
		}

		const leftHeight = dfs(node.left);
		const rightHeight = dfs(node.right);
		res = Math.max(res, leftHeight + rightHeight);
		return 1 + Math.max(leftHeight, rightHeight);
	}
}

File: neetcode-250/trees/543-diameter-of-binary-tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return (-1, -1)

            left_hei, left_diam = dfs(node.left)
            right_hei, right_diam = dfs(node.right)
            return (
                max(left_hei, right_hei) + 1,
                max(left_diam, right_diam, left_hei + right_hei + 2),
            )

        return max(dfs(root))
